packages/nimble-components/src/table-column/duration-text/models/duration-formatter.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * A class for formatting a numeric value in the unit of seconds into a display value representing its
+ *  days, hours, minutes, and seconds.
+ */
+export class DurationFormatter {
+    private readonly daysFormatter: Intl.NumberFormat;
+    private readonly hoursFormatter: Intl.NumberFormat;
+    private readonly minutesFormatter: Intl.NumberFormat;
+    private readonly secondsFormatter: Intl.NumberFormat;
+    private readonly scientificSecondsFormatter: Intl.NumberFormat;
+
+    public constructor(private readonly lang: string) {
+        this.daysFormatter = new Intl.NumberFormat(this.lang, {
+            style: 'unit',
+            unit: 'day'
+        });
+        this.hoursFormatter = new Intl.NumberFormat(this.lang, {
+            style: 'unit',
+            unit: 'hour'
+        });
+        this.minutesFormatter = new Intl.NumberFormat(this.lang, {
+            style: 'unit',
+            unit: 'minute'
+        });
+        this.secondsFormatter = new Intl.NumberFormat(this.lang, {
+            style: 'unit',
+            unit: 'second'
+        });
+        this.scientificSecondsFormatter = new Intl.NumberFormat(this.lang, {
+            style: 'unit',
+            unit: 'second',
+            notation: 'scientific',
+            maximumFractionDigits: 3
+        });
+    }
+
+    public format(value: number | null | undefined): string {
+        if (
+            value === null
+            || value === undefined
+            || value < 0
+            || !Number.isFinite(value)
+        ) {
+            return '';
+        }
+
+        const result = [];
+        const fractionalDays = value / 86400000;
+        let remainingTime = value;
+        const days = Math.floor(fractionalDays);
+        if (days <= 100 && days > 0) {
+            const formattedDays = this.daysFormatter.format(days);
+            result.push(formattedDays);
+            remainingTime -= 86400000;
+        } else if (days >= 100) {
+            return this.scientificSecondsFormatter.format(value / 1000);
+        }
+
+        const hours = Math.floor((value / 3600000) % 24);
+        if (hours) {
+            const formattedHours = this.hoursFormatter.format(hours);
+            result.push(formattedHours);
+            remainingTime -= 3600000;
+        }
+
+        const minutes = Math.floor((value / 60000) % 60);
+        if (minutes) {
+            const formattedMinutes = this.minutesFormatter.format(minutes);
+            result.push(formattedMinutes);
+            remainingTime -= 60000;
+        }
+
+        const valueInSeconds = remainingTime / 1000;
+        const seconds = valueInSeconds === 0 ? 0 : valueInSeconds % 60;
+        if (value < 1 && valueInSeconds !== 0 && result.length === 0) {
```

Comment:
Not quite. So, for starters, `value` can both be less than 1 _and_ 0, and if it _is_ 0, we don't want to use the scientific formatter. That said, there _is_ an unnecessary check here, which is the `result.length === 0` check, which will _always_ be true when `value < 1` (by virtue of the preceding logic in the function). Note that we want to use the `valueInSeconds !== 0` check instead of `milliseconds !== 0` to handle the case when `milliseconds` is actually a really small value that results in a rounded value of `0` for `valueInSeconds`.

Reply:
I'm not sure why I wasn't considering the possibility that `value` was 0. Makes sense.